src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+    // created for sb-edit), "inputs" are stored in not one but two containers
+    // per block: inputs, and fields. The difference is discussed in their
+    // corresponding functions. Blocks may also carry a mutation, a mapping of
+    // XML property names and values, for use in some blocks (notably those
+    // associated with custom blocks). All this data is serialized and detailed
+    // in serializeInputs.
+    //
+    // serializeBlock is in charge of serializing an individual block, as well
+    // as its following block, and building the links between it and its parent
+    // and siblings. As with other block-related functions, data is collected
+    // into a BlockData object, a wrapper around the mapping of IDs to their
+    // associated serialized block.
+    //
+    // It's possible for a block to be skipped altogether during serialization,
+    // because it referred to some value which could not be converted into
+    // valid SB3 data. For reporters, this means leaving an empty input; for
+    // stack blocks, it means skipping to the next block in the sibling array
+    // (or leaving an empty connection if there is none). It's up to the caller
+    // to handle serializeBlock returning a null blockId usefully.
+
+    const blockData = newBlockData();
+
+    const { getBroadcastId, getCustomBlockData, parent, siblingBlocks, stage, target } = options;
+
+    let nextBlock: Block;
+    let nextBlockId: sb3.Block["next"];
+    if (siblingBlocks) {
+      const thisIndex = siblingBlocks.indexOf(block) + 1;
+      nextBlock = siblingBlocks.find((x, index) => index === thisIndex);
+    }
+
+    if (nextBlock) {
+      const { blockData: nextBlockData, blockId } = serializeBlock(nextBlock, {
+        stage,
+        target,
+        getBroadcastId,
+        getCustomBlockData,
+        parent: block,
+        siblingBlocks
+      });
+
+      applyBlockData(blockData, nextBlockData);
+      nextBlockId = blockId;
+    }
+
+    const serializeInputsResult = serializeInputs(block, {
+      stage,
+      target,
+
+      getBroadcastId,
+      getCustomBlockData
+    });
+
+    if (!serializeInputsResult) {
+      return { blockData, blockId: nextBlockId };
+    }
+
+    const { inputs, fields, mutation, blockData: inputBlockData } = serializeInputsResult;
+
+    applyBlockData(blockData, inputBlockData);
+
+    const obj: sb3.Block = {
+      opcode: block.opcode,
+
+      parent: parent ? parent.id : null,
+      next: nextBlockId,
+      topLevel: !parent,
+
+      inputs,
+      fields,
+      mutation,
+
+      shadow: false
+    };
+
+    if (obj.topLevel) {
+      obj.x = options.x;
+      obj.y = options.y;
+    }
+
+    const blockId = block.id;
+
+    blockData[blockId] = obj;
+
+    return { blockData, blockId };
+  }
+
+  interface CustomBlockData {
+    args: Array<{
+      default: string;
+      id: string;
+      name: string;
+      type: "boolean" | "numberOrString";
+    }>;
```

Comment:
Could
```ts
    {
      default: string;
      id: string;
      name: string;
      type: "boolean" | "numberOrString";
    }
```

be made into a `CustomBlockArg` interface so that `args` below can be an `Array<CustomBlockArg>`?